optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/WeightedScoreImpacter.java

```diff
@@ -16,6 +16,78 @@
 
 package org.optaplanner.core.impl.score.inliner;
 
+import java.math.BigDecimal;
+import java.util.function.BiFunction;
+
+/**
+ * There are several valid ways how an impacter could be called from a constraint stream:
+ *
+ * <ul>
+ * <li>{@code .penalize(..., (int) 1)}</li>
+ * <li>{@code .penalizeLong(..., (int) 1)}</li>
+ * <li>{@code .penalizeLong(..., (long) 1)}</li>
+ * <li>{@code .penalizeBigDecimal(..., (int) 1)}</li>
+ * <li>{@code .penalizeBigDecimal(..., (long) 1)}</li>
+ * <li>{@code .penalizeBigDecimal(..., BigDecimal.ONE)}</li>
+ * <li>Plus reward variants of the above.</li>
+ * </ul>
+ *
+ * Implementations of this class need to be able to handle all of these, casting to the appropriate type.
+ * In certain cases, such as when BigDecimal is passed into an int impacter, methods are allowed to throw exceptions.
+ * That state should be considered impossible, ruled out by the CS API itself.
+ */
 public interface WeightedScoreImpacter {
 
+    static WeightedScoreImpacter of(IntImpactFunction impactFunction) {
+        return new IntWeightedScoreImpacter(impactFunction);
+    }
+
+    static WeightedScoreImpacter of(LongImpactFunction impactFunction) {
+        return new LongWeightedScoreImpacter(impactFunction);
+    }
+
+    static WeightedScoreImpacter of(BigDecimalImpactFunction impactFunction) {
+        return new BigDecimalWeightedScoreImpacter(impactFunction);
+    }
+
+    /**
+     * @param matchWeight never null
+     * @param justificationsSupplier never null
+     * @return never null
+     */
+    UndoScoreImpacter impactScore(int matchWeight, JustificationsSupplier justificationsSupplier);
+
+    /**
+     * @param matchWeight never null
+     * @param justificationsSupplier never null
+     * @return never null
+     */
+    UndoScoreImpacter impactScore(long matchWeight, JustificationsSupplier justificationsSupplier);
+
+    /**
+     * @param matchWeight never null
+     * @param justificationsSupplier never null
+     * @return never null
+     */
+    UndoScoreImpacter impactScore(BigDecimal matchWeight, JustificationsSupplier justificationsSupplier);
+
+    @FunctionalInterface
+    interface IntImpactFunction {
+
+        UndoScoreImpacter impact(int matchWeight, JustificationsSupplier justificationsSupplier);
+
+    }
+
+    @FunctionalInterface
+    interface LongImpactFunction {
+
+        UndoScoreImpacter impact(long matchWeight, JustificationsSupplier justificationsSupplier);
+
+    }
+
+    @FunctionalInterface
+    interface BigDecimalImpactFunction extends BiFunction<BigDecimal, JustificationsSupplier, UndoScoreImpacter> {
```

Comment:
Its a bit clunky that the other interfaces use "impact" as a method name and this one uses "apply"

Reply:
Java has ToIntBiFunction, but nothing that takes an <int,A> returns B.
So I'd just not inherit BiFunction here and just declare it similar to the others?

(soft suggestion)